sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/JwtSignedBase.cs

```diff
@@ -0,0 +1,79 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis;
+using System;
+using SonarAnalyzer.Helpers;
+using System.Collections.Immutable;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class JwtSignedBase<TSyntaxKind, TInvocationSyntax> : SonarDiagnosticAnalyzer
+        where TSyntaxKind : struct
+        where TInvocationSyntax : SyntaxNode
+    {
+        protected const string DiagnosticId = "S5659";
+        private const string MessageFormat = "Use only strong cipher algorithms when {0} this JWT.";
+        private const string MessageVerifying = "verifying the signature of";
+        protected const bool JwtBuilderConstructorIsSafe = false;
+
+        protected readonly DiagnosticDescriptor verifyingRule;
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(verifyingRule);
+
+        protected InvocationTracker<TSyntaxKind> InvocationTracker { get; set; }
+
+        protected abstract BuilderPatternCondition<TInvocationSyntax> BuilderPattern();
+
+        protected JwtSignedBase(System.Resources.ResourceManager rspecResources)
+        {
+            verifyingRule = DiagnosticDescriptorBuilder
+                .GetDescriptor(DiagnosticId, string.Format(MessageFormat, MessageVerifying), rspecResources)
+                .WithNotConfigurable();
+        }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            InvocationTracker.Track(context,
+                InvocationTracker.MatchMethod(
+                    new MemberDescriptor(KnownType.JWT_IJwtDecoder, "Decode"),
+                    new MemberDescriptor(KnownType.JWT_IJwtDecoder, "DecodeToObject")),
+                Conditions.Or(
+                    InvocationTracker.ArgumentIsBoolConstant("verify", false),
+                    InvocationTracker.MethodHasParameters(1)
+                    ));
+
+            InvocationTracker.Track(context,
+                InvocationTracker.MatchMethod(new MemberDescriptor(KnownType.JWT_Builder_JwtBuilder, "Decode")),
+                InvocationTracker.IsInvalidBuilderInitialization(BuilderPattern())
+                );
+        }
+
+        protected BuilderPatternDescriptor<TInvocationSyntax>[] JwtBuilderDescriptors(Func<InvocationContext, TInvocationSyntax, bool> singleArgumentIsNotFalseLiteral)
```

Comment:
It's not used in this rule, but it's there for reusability purposes. Future usages might need to know what `context.Invocation` started this tracking to make correct decision.

Reply:
This is just an assumption currently we don't know yet for sure if this will be needed or not. I suggest to remove this now and add it later when needed.